### PR TITLE
Add Op2d->add_const collapse substitution

### PR DESCRIPTION
**Pull request by elad-c**
## Pull Request Description:
Add a substitution for collapsing Op2d->add-const into the Op2d node
Op2d: Conv2D, DW-Conv2D, Conv2DTranspose or Dense

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

**Comment by eladc-git** [comment]
General question: why implementation is per framework? if I understand it correctly, the substitution is only adding the const to Op2D_bias so it can be done on graph (common). I didn't see the "per framework" factor here.

File: model_compression_toolkit/core/common/substitutions/linear_collapsing_substitution.py [hunk 1]
```diff
@@ -30,6 +30,9 @@ def linear_collapsing_substitute(graph: common.Graph,
     Returns:
         Transformed graph after applying all linear collapsing substitutions.
     """
+    # TODO: remove this if after adding Op2d-add_const collapse substitution in PyTorch
+    if linear_collapsing_substitution is None:
```
Comment (eladc-git) [inline model_compression_toolkit/core/common/substitutions/linear_collapsing_substitution.py:34]: what is this TODO? and why pytorch is mentioned in common?

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/linear_collapsing.py [hunk 2]
```diff
@@ -123,3 +127,69 @@ def keras_linear_collapsing() -> Conv2DCollapsing:
...
+    """
+    bias = op2d_node.get_weights_by_keys(bias_str)
+
+    # read constant from add node
+    if len(add_node.op_call_args) > 0:
+        const = add_node.op_call_args[0]
+    elif 'y' in add_node.op_call_kwargs:
+        const = add_node.op_call_kwargs['y']
+    else:
+        Logger.error(f'Unable to read constant from add node: {add_node.name}')
```
Comment (eladc-git) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/linear_collapsing.py:167]: add no cover

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/linear_collapsing.py [hunk 3]
```diff
@@ -123,3 +127,69 @@ def keras_linear_collapsing() -> Conv2DCollapsing:
...
+
+    # convert constant to numpy array
+    if isinstance(const, tf.Tensor):
+        const = const.numpy()
+    elif isinstance(const, list):
+        const = np.array(const)
+    else:
+        Logger.error(f'Unable to convert constant to numpy array: {add_node.name}')
```
Comment (eladc-git) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/linear_collapsing.py:175]: no cover

File: model_compression_toolkit/core/pytorch/pytorch_implementation.py [hunk 4]
```diff
@@ -289,6 +289,12 @@ def get_linear_collapsing_substitution(self) -> common.BaseSubstitution:
         """
         return pytorch_linear_collapsing()
 
+    def get_op2d_add_const_collapsing_substitution(self) -> common.BaseSubstitution:
+        """
+        Returns: None, as Op2d add-const substitution is not supported in torch yet
+        """
+        return None
```
Comment (eladc-git) [inline model_compression_toolkit/core/pytorch/pytorch_implementation.py:296]: what about pytorch implementation?

File: tests/keras_tests/feature_networks_tests/feature_networks/linear_collapsing_test.py [hunk 5]
```diff
@@ -125,5 +125,107 @@ def create_networks(self):
...
+        x2 = layers.Dense(9, use_bias=True, bias_initializer='glorot_uniform')(x)
+        x = tf.math.add(x2, x)
+        y = layers.ReLU()(x)
+
+        return tf.keras.models.Model(inputs=inputs, outputs=y)
+
+    def compare(self, quantized_model, float_model, input_x=None, quantization_info=None):
+        super().compare(quantized_model, float_model, input_x, quantization_info)
+        for layer in quantized_model.layers:
+            if type(layer) in [layers.Conv2D, layers.DepthwiseConv2D, layers.Conv2DTranspose, layers.Dense]:
```
Comment (eladc-git) [inline tests/keras_tests/feature_networks_tests/feature_networks/linear_collapsing_test.py:230]: you can save the const in the init function and then check here if it equals to the bias (when bias is none).

**Review by eladc-git — CHANGES_REQUESTED** [review]

**Review by eladc-git — APPROVED** [review]

File: model_compression_toolkit/core/common/substitutions/linear_collapsing.py [hunk 6]
```diff
@@ -182,3 +179,83 @@ def substitute(self,
...
+        # If there is an extra connection between these two nodes skip the substitution
+        if len(graph.get_next_nodes(first_node)) > 1 or len(graph.get_prev_nodes(second_node)) > 1:
+            return graph
+
+        # New collapsed bias
+        bias = self.op2d_collapsing_fn(first_node, second_node, self.bias_str)
+
+        # New collapsed node
+        op2d_collapsed = copy.deepcopy(first_node)
+        op2d_collapsed_name = first_node.name + '_collapsed'
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/substitutions/linear_collapsing.py:243]: I think the name manipulation is meaningless and may add confusion... what do you think?

**Review by reuvenperetz — APPROVED** [review]
